### PR TITLE
TECH-5529: skip storing bitfield keys if default

### PR DESCRIPTION
**Pull request by ttstarck**
Bitfields are currently writing their keys in `to_store` even for default values. This is a small change to stop that from happening.

### CODE REVIEW

**Comment by ttstarck** [comment]
I'll update the CHANGELOG and versions shortly

File: lib/aggregate/attribute/bitfield.rb [hunk 1]
```diff
@@ -40,6 +40,10 @@ def validation_errors(_value)
     []
   end
 
+  def skip_default?
```
Comment (alecjacobs5401) [inline lib/aggregate/attribute/bitfield.rb:43]: the name of this method leaves me with lots of questions. It isn't really clear what we are skipping the default for.

Side note - do we always want to skip the default? Or do we only care about skipping the default when the default is actually `nil`?

File: lib/aggregate/aggregate_store.rb [hunk 2]
```diff
@@ -119,7 +119,7 @@ def to_store
 
         # Optimization: only write out values if they are not nil, if there is no schema migration and
         # the default value is nil. (Schema migrations and defaults depend on writing the value.)
-        if respond_to?(:data_schema_version) || !aa.default.nil? || !agg_value.nil?
```
Comment (ttstarck) [inline lib/aggregate/aggregate_store.rb:122]: @primiti the ticket body, you state that you thought we weren't writing default values. However this currently only doesn't write the default if the default is nil and the value is nil.

**Review by alecjacobs5401 — CHANGES_REQUESTED** [review]
This change feels a bit like we're breaking Liskov Substitution.

We're adding a method to a single child class and we have to account for that in another process by using `respond_to?`.

We can accomplish a similar thing by inverting the question we added and creating a method on the base class that could be
```rb
store_default_value?
```

Which can be overwritten as necessary. 

An even simpler approach would be to see if we can always skip writing if the value matches the defined default. That may have longer term implications though, as discussed offline previously.

What do you think about that?

File: lib/aggregate/attribute/bitfield.rb (hunk 1 above)
Comment (primiti) [reply lib/aggregate/attribute/bitfield.rb:43]: 'do_not_write_default_values?' is closest to its current form, but I would invert it.  'write_default_values?' is easier to understand.

File: lib/aggregate/aggregate_store.rb (hunk 2 above)
Comment (primiti) [reply lib/aggregate/aggregate_store.rb:122]: Aha!  That could be the issue.  So the question is what breaks if we change this?  Can we try this and see if anything breaks in web?

File: lib/aggregate/aggregate_store.rb [hunk 3]
```diff
@@ -285,5 +285,13 @@ def ensure_saved_changes_up_to_date
         save_changes
       end
     end
+
+    def store_value?(aggregate_attribute, value)
+      skip_on_default =
+        if aggregate_attribute.respond_to?(:skip_default?)
```
Comment (primiti) [inline lib/aggregate/aggregate_store.rb:291]: You can add this to the base class and then we don't need to check respond_to?  It would be clearer that way as well.

File: test/dummy/test/unit/passport_test.rb [hunk 4]
```diff
@@ -72,7 +72,6 @@ class PassportTest < ActiveSupport::TestCase
           "city"           => "Santa Barbara",
           "foreign_visits" => [{ "country" => "Canada" }, { "country" => "Mexico" }],
           "gender"         => "female",
-          "stamps"         => nil,
           "state"          => "California",
           "weight"         => "100.0"
         }
```
Comment (primiti) [inline test/dummy/test/unit/passport_test.rb:77]: Can you add a test that shows the bitfield not being stored when empty?

**Review by primiti — CHANGES_REQUESTED** [review]
I think you found the core problem and your solution is about 90% of the way there.  Can we go a little further?  

I suggested inverting the meaning and renamed your `skip_default?` method and renamed it `write_default_values?`.   Can we remove this from bitfield and add it to **Aggregate::Attribute::Base** as a new option that defaults to true?   

Then can we then integrate with web and see if anything breaks.  If any of the fields need to be written at default values we can add `write_default_values: true` to the field declaration.   

I am pretty sure this will make a very big difference in the average size of calls.
